pkg/live/inventoryrg.go

```diff
@@ -249,78 +308,432 @@ func (icm *InventoryResourceGroup) GetObject() (*unstructured.Unstructured, erro
+// collapse is a utility function to rebalance objects stored within sharded ResourceGroups.
+func (icm *InventoryResourceGroup) collapse(dc dynamic.Interface, mapper meta.RESTMapper) error {
+	if ResourceCount == DisableSharding {
+		klog.V(4).Info("skipping re-balancing of ResourceGroups as sharding is disabled")
+		return nil
+	}
+
+	namespacedClient, err := icm.getNamespacedClient(dc, mapper)
 	if err != nil {
 		return err
 	}
 
-	// Update the cluster inventory object.
-	appliedObj, err := namespacedClient.Update(context.TODO(), invInfo, metav1.UpdateOptions{})
+	// Get all cluster ResourceGroup objects using label selector.
+	resourceList, err := namespacedClient.List(context.TODO(), metav1.ListOptions{
+		LabelSelector: fmt.Sprintf(shardedRGLabelFmt, icm.Name()),
+	})
+	if err != nil {
+		return err
+	}
+
+	resourcegroups := resourceList.Items
+	sortResourceGroups(resourcegroups)
+
+	// Get all GKNNs across all sharded ResourceGroups.
+	items := []interface{}{}
+	for _, rg := range resourcegroups {
+		resources, exists, err := unstructured.NestedSlice(rg.Object, "spec", "resources")
+		if err != nil || !exists {
+			continue
+		}
 
-	// Update status.
-	if statusPolicy == inventory.StatusPolicyAll {
-		invInfo.SetResourceVersion(appliedObj.GetResourceVersion())
-		_, err = namespacedClient.UpdateStatus(context.TODO(), invInfo, metav1.UpdateOptions{})
+		items = append(items, resources...)
 	}
 
-	return err
+	count := len(items)
+
+	// Re-balance GKNN objects in sharded ResourceGroups.
+	objIdx := 0
+	for objIdx < count {
+		var curr []interface{}
+		if objIdx+ResourceCount < len(items) {
+			curr = items[objIdx : objIdx+ResourceCount]
+		} else {
+			curr = items[objIdx:]
+		}
+		err := unstructured.SetNestedSlice(resourcegroups[objIdx/ResourceCount].UnstructuredContent(), curr, "spec", "resources")
+		if err != nil {
+			return err
+		}
+		_, err = namespacedClient.Update(context.TODO(), &resourcegroups[objIdx/ResourceCount], metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+		objIdx += ResourceCount
+	}
+
+	// Delete ResourceGroups that contain duplicated GKNNs.
+	for rgIdx := len(resourcegroups) - 1; rgIdx >= objIdx/ResourceCount; rgIdx-- {
+		err = namespacedClient.Delete(context.TODO(), resourcegroups[rgIdx].GetName(), metav1.DeleteOptions{})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// createResourceGroups creates new ResourceGroup object(s) on cluster, and handles the sharding of managed
+// resources into multiple ResourceGroup objects.
+func (icm *InventoryResourceGroup) createResourceGroups(dc dynamic.Interface, mapper meta.RESTMapper) error {
+	namespacedClient, err := icm.getNamespacedClient(dc, mapper)
+	if err != nil {
+		return err
+	}
+
+	// totalStored tracks the total number of resources stored within all sharded ResourceGroups.
+	totalStored := 0
+
+	for totalStored < len(icm.objMetas) {
+		var items []interface{}
+
+		switch ResourceCount {
+		case DisableSharding:
```

Comment:
Not really loving this disabled = -1 as an enum pattern. Maybe just use `ResourceCount > 0` for enabled and <= 0 for disabled?

Reply:
Refactored to have resourceCount as an internal variable, with a new method `shardingDisabled()` to check `<=0` for disabled.